### PR TITLE
feature/tests-setup

### LINKED ISSUE #11: Research Electron testing practices
Due to the fact that we want to (at least try to) write cleaner code we need to write tests.
Goals is to research & review general JS/TS testing approaches and frameworks. Then look specifically for testing Electron apps.

• You might take a look at TDD, BDD - maybe we will decide to adopt them if we feel so.
• Also you might take a look at Snapshot testing and whether or not it is useful for Electron and how to do it (if it is possible)

• https://www.electronjs.org/docs/development/testing
• https://github.com/mawrkus/js-unit-testing-guide
• https://jestjs.io/
• https://mochajs.org/
• https://www.chaijs.com/

### PR DESCRIPTION
**Pull request by yerkeboolean**
closes #11 

Check this [issue](https://github.com/crococoders/tchaikovsky/issues/11) to know which testing technologies are used and related info.

[Here ](https://github.com/crococoders/tchaikovsky/blob/feature/tests-setup/howToTest.md) you can find instructions how to write Unit and Snapshot tests for components

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -15,5 +15,5 @@
     "noEmit": true,
     "jsx": "react"
   },
-  "include": ["src"]
-}
\ No newline at end of file
+  "include": ["src", "src/setupTests.ts"]
```
Comment (aidarnouman) [inline tsconfig.json]: I think this one is redundant, because `src/` is already included. Correct me if I am wrong

File: src/setupTests.ts [hunk 2]
```diff
@@ -0,0 +1,4 @@
+import { configure } from 'enzyme';
```
Comment (aidarnouman) [inline src/setupTests.ts]: Let's create new directory called `tests/` nearby `src/` and put this file there (you can rename it as just `setup.ts`) as well.

File: howToTest.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
+[Jest](https://jestjs.io/docs/en/expect) main testing framework
```
Comment (aidarnouman) [inline howToTest.md]: Can you add a linebreak after this line? 
Because in the preview both **Jest** and **Enzyme** are on the same line

File: howToTest.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+[Enzyme](https://enzymejs.github.io/enzyme) - Helper testing framework for getting elements from the DOM
+
+### Basic settings:
+
+1. Create '**tests**' folder in component's folder
+2. Create <my-component>.spec.tsx
+3. First line must be (it's IMPORTANT) **import 'jsdom-global/register';**
+4. Then import :
+   - import React from 'react';
+   - import { MyCompnent } from '../MyCompnent'; // your components
```
Comment (aidarnouman) [inline howToTest.md]: I think you can group bullets 3 and 4 into one inside one code-block, like:

```javascript
import 'jsdom-global/register' // this one should be first, it's important!

import React from 'react'
import { MyComponent } from '../MyComponent'
```

File: howToTest.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
+[Jest](https://jestjs.io/docs/en/expect) main testing framework
+[Enzyme](https://enzymejs.github.io/enzyme) - Helper testing framework for getting elements from the DOM
+
+### Basic settings:
+
+1. Create '**tests**' folder in component's folder
+2. Create <my-component>.spec.tsx
```
Comment (aidarnouman) [inline howToTest.md]: `spec` is the nice word, but it has generally more meaning in the world of software engineering and it is less obvious than just `test`. 
I suggest to make it `<my-component>.test.tsx`, what do you think?

File: howToTest.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+3. First line must be (it's IMPORTANT) **import 'jsdom-global/register';**
+4. Then import :
+   - import React from 'react';
+   - import { MyCompnent } from '../MyCompnent'; // your components
+
+### Unit testing of component:
+
+1.  import { shallow } from 'enzyme';
```
Comment (aidarnouman) [inline howToTest.md]: Please put every piece of code inside specified code-block:  \`\`\`javascript \`\`\`

File: howToTest.md [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+     shallow or mount or render
+     You should read enzyme [API reference](https://enzymejs.github.io/enzyme/docs/api/) to
+
+2. write your test using 'test' or 'it' function:
```
Comment (aidarnouman) [inline howToTest.md]: You should write here about grouping our tests with `describe()` and how should we name our tests. Because naming the tests is **really** important, consider the following example:

```javascript
describe('Testing the appController', () => {
    it('should add two numbers together', () => {
      
    });

    it('should substract the second number from the first one', () => {
      
    });

    it('should multiply the two numbers together', () => {
        
    });

    it('should divide the first number with the second', () => {
        
    });
});
```

Notice how individual testcase is being read as plain English sentence - this is good naming. Also I suggest to stick with `it()` and do not use `test()` at all.

[Check it out](https://medium.com/javascript-in-plain-english/writing-your-first-unit-test-with-jest-b4953dabb931)

P.S. Also your sentence needs to be capitalized: _2. **W**rite your test using 'test' or 'it' function:_

File: howToTest.md (hunk 5 above)
Comment (aidarnouman) [reply howToTest.md]: It basically depends on whether we are following BDD or TDD to write our tests. [Check this article](https://www.bignerdranch.com/blog/why-do-javascript-test-frameworks-use-describe-and-beforeeach/)
You need to decide on the naming `spec` or `test` after you decide which testing approach will we adopt (TDD or BDD)

File: howToTest.md [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+3. Write body of your test:
+
+```javascript
+const wrapper = shallow(<MyCompnent title="Hello Jest!" />); // get your component and assign to variable
+expect(wrapper.html()).toMatch('<div>Hello Jest!</div>');
+// check for eqality(or match) you component output with your expected data
+```
+
+4. type 'npm test' or 'npm run test' in command line
```
Comment (aidarnouman) [inline howToTest.md]: - Commands also better be wrapped inside \`\`\`code blocks \`\`\`
- Sentence must be capitalized

**Review by nugmanoff — CHANGES_REQUESTED** [review]

**Comment by nugmanoff** [comment]
You have some recurring mistakes in `howToTest.md` (like capitalization, code-blocks and etc.)
I didn't mention all of them, but please make sure to double-check the whole document.

**Comment by nugmanoff** [comment]
Also you can rename `howToTest.md` to `testing.md` and move it to separate `docs/` folder. (Please put `development.md` there too).

**Comment by nugmanoff** [comment]
That's why we need commit linting :)
It should have been: `docs: update howToTest.md`

Screen Shot 2020-04-20 at 20 06 15

**Comment by nugmanoff** [comment]
As you promised it is quite solid PR! Good job, brother!💪🏻
I really liked the documentation & examples.

File: tsconfig.json (hunk 1 above)
Comment (yerkeboolean) [reply tsconfig.json]: it was created automatically. I thought that was required to include it.

**Review by nugmanoff — APPROVED** [review]

File: tsconfig.json [hunk 9]
```diff
@@ -5,15 +5,16 @@
     "allowJs": true,
     "skipLibCheck": true,
     "esModuleInterop": true,
+    "experimentalDecorators": true,
     "allowSyntheticDefaultImports": true,
     "strict": true,
     "forceConsistentCasingInFileNames": true,
-    "module": "esnext",
+    "module": "commonjs",
```
Comment (aidarnouman) [inline tsconfig.json]: Why did you change this one?

File: package.json [hunk 10]
```diff
@@ -70,23 +75,54 @@
...
     "electron": "8.2.1",
+    "enzyme": "^3.11.0",
+    "enzyme-adapter-react-16": "^1.15.2",
+    "enzyme-to-json": "^3.4.4",
     "eslint": "^6.8.0",
     "eslint-plugin-import": "^2.20.0",
     "fork-ts-checker-webpack-plugin": "^3.1.1",
+    "jest": "^25.3.0",
+    "jest-cli": "^25.3.0",
+    "jest-cucumber": "^2.0.11",
```
Comment (aidarnouman) [inline package.json]: Do we need this one here?

File: bdd/step_definitions/ExampleSteps.ts [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+    this.today = 'Sunday';
+  }
+
+  @when(/I ask whether it's Friday yet/)
+  public whenAskIsItFriday() {
+    this.actualAnswer = this.isItFriday(this.today);
+  }
+
+  @then(/I should be told "([^"]*)"/)
+  public accountBalanceShouldEqual(expectedAnswer: string) {
```
Comment (aidarnouman) [inline bdd/step_definitions/ExampleSteps.ts]: I think you misnamed the function

File: cucumber.js [hunk 12]
```diff
@@ -0,0 +1,11 @@
+let common = [
+  'bdd/features/**/*.feature', // Specify our feature files
```
Comment (aidarnouman) [inline cucumber.js]: Can we do this way:
- Rename `bdd/` to `tests/` (because it is really not necessary to show what kind of testing technique we are using  - be it BDD or TDD)
- Move `srcs/tests/setup.ts` to new directory `tests/`
- Rename `step_definitions/` folder to just `steps/`, because it is redundant I think
- Rename `ExampleSteps.ts` to just `Example.ts`, because it is redundant too

**Comment by nugmanoff** [comment]
I think it's good to go, just a few finishing touches - also I will review the formatting & grammar of `testing.md` after you finished (I'll do the fixes myself, don't worry)

File: tsconfig.json (hunk 9 above)
Comment (yerkeboolean) [reply tsconfig.json]: When I use `esnext` it throws an error image
so I decided to change it to `commonjs`, so it works

File: package.json (hunk 10 above)
Comment (yerkeboolean) [reply package.json]: We need jest and jest-cli, but don't need jest-cucumber so I delete it

File: tsconfig.json (hunk 9 above)
Comment (aidarnouman) [reply tsconfig.json]: So it is quite 'hacky' solution, whenever you introduce a change - you really need to understand why are you doing it. Especially when you are changing the project-wide tsconfig file.

[Read this](https://stackoverflow.com/questions/39883960/what-is-the-typescript-2-0-es2015-way-to-import-assert-from-node-js) or we can switch over to [chai](https://www.chaijs.com/) (actually we already have it in the project)

[This one](https://lihautan.com/javascript-modules/) is a wonderful article about Javascript modules history

**Review by nugmanoff — APPROVED** [review]
